hack.html.markdown

```diff
@@ -1,308 +1,381 @@
-class ConsistentFoo
-{
-    public function __construct(int $x, float $y)
-    {
-        // ...
+    public function get(): int {
+      return $this->i;
     }
+  }
 
-    public function someMethod()
-    {
-        // ...
+  // Properties and Methods can be static (not requiring instantiation).
+  class Person {
+    public static function favoriteProgrammingLanguage(): string {
+      return "Hack";
     }
-}
+  }
 
-class ConsistentBar extends ConsistentFoo
-{
-    public function __construct(int $x, float $y)
-    {
-        // Hack's type checker enforces that parent constructors are called
-        parent::__construct($x, $y);
+  function demo_hack_classes(): void {
+    // Use `new` to instantiate a class.
+    $c1 = new Counter();
 
-        // ...
-    }
+    // To call a static property or method, use `::`
+    $typical_person = tuple("Andrew", Person::favoriteProgrammingLanguage());
+  }
 
-    // The __Override annotation is an optional signal for the Hack type
-    // checker to enforce that this method is overriding a method in a parent
-    // or trait. If not, this will error.
-    <<__Override>>
-    public function someMethod()
-    {
-        // ...
+  // Abstract class can be defined, but not instantiated directly.
+  abstract class Machine {
+    public function openDoors(): void {
+      return;
     }
-}
-
-class InvalidFooSubclass extends ConsistentFoo
-{
-    // Not matching the parent constructor will cause a type checker error:
-    //
-    //  "This object is of type ConsistentBaz. It is incompatible with this object
-    //   of type ConsistentFoo because some of their methods are incompatible"
-    //
-    public function __construct(float $x)
-    {
-        // ...
+    public function closeDoors(): void {
+      return;
     }
-
-    // Using the __Override annotation on a non-overridden method will cause a
-    // type checker error:
-    //
-    //  "InvalidFooSubclass::otherMethod() is marked as override; no non-private
-    //   parent definition found or overridden parent is defined in non-<?hh code"
-    //
-    <<__Override>>
-    public function otherMethod()
-    {
-        // ...
+  }
+
+  /* ==================================
+   *             INTERFACES
+   * ==================================
+   */
+
+  // A class can implement a set of capabilities via an interface.
+  // An interface is a set of method declarations and constants.
+
+  interface Plane {
+    // A constant is a named value. Once defined, the value cannot be changed.
+    const MAX_SPEED = 300;
+    public function fly(): void;
+  }
+
+  /* ==================================
+   *             TRAITS
+   * ==================================
+   */
+
+  // A trait defines properties and method declarations.
+  // Traits are recommended when abstracting code for reuse.
+  // Traits are included in code via the `use` keyword.
+  // `use` allows for other includes, like namespaces, classes, and functions (and more)!
```

Comment:
What do you mean with _allows for other includes_?
Inclusion directives (`require_once`, `include`, etc.).

Reply:
Now *I'm* wondering what I meant. Removing because `use` can only be used with Traits (right?).

I'm removing this line because I was likely conflating ideas (and tired).